csharp/src/Google.Protobuf/Collections/RepeatedFieldExtensions.cs

```diff
@@ -0,0 +1,53 @@
+#region Copyright notice and license
+// Protocol Buffers - Google's data interchange format
+// Copyright 2015 Google Inc.  All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd
+#endregion
+
+using System;
+
+namespace Google.Protobuf.Collections
+{
+    /// <summary>
+    /// A set of extension methods on RepeatedField
+    /// </summary>
+    public static class RepeatedFieldExtensions
+    {
+        /// <summary>Adds the elements of the specified span to the end of the <see cref="RepeatedField{T}"/>.</summary>
+        /// <typeparam name="T">The type of elements in the <see cref="RepeatedField{T}"/>.</typeparam>
+        /// <param name="repeatedField">The list to which the elements should be added.</param>
+        /// <param name="source">The span whose elements should be added to the end of the <see cref="RepeatedField{T}"/>.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="repeatedField"/> is null.</exception>
+        public static void AddRange<T>(this RepeatedField<T> repeatedField, ReadOnlySpan<T> source)
+        {
+            ProtoPreconditions.CheckNotNull(repeatedField, nameof(repeatedField));
+
+            if (source.IsEmpty)
+            {
+                return;
+            }
+
+            // For reference types and nullable value types, we need to check that there are no nulls
+            // present. (This isn't a thread-safe approach, but we don't advertise this is thread-safe.)
+            // We expect the JITter to optimize this test to true/false, so it's effectively conditional
+            // specialization.
+            if (default(T) == null)
+            {
+                foreach (var value in source)
+                {
+                    if (value == null)
+                    {
+                        throw new ArgumentException("ReadOnlySpan contained null element", nameof(source));
+                    }
+                }
+            }
```

Comment:
I'm sure there must be a faster way to check if there is a null in a span.

Does this work? IndexOf must have SIMD optimizations to quickly search the span.
```suggestion
            if (default(T) == null)
            {
                if (source.IndexOf(null) != -1)
                {
                    throw new ArgumentException("ReadOnlySpan contained null element", nameof(source));
                }
            }
```

Reply:
The latest code uses a for loop:

```csharp
for (int i = 0; i < source.Length; i++)
                {
                    if (source[i] == null)
                    {
                        throw new ArgumentException("ReadOnlySpan contained null element", nameof(source));
                    }
                }
```

Reply:
More context:

source here is a ReadOnlySpan, so just foreach'ing it will be lowered by the C# compiler into the equivalent for loop.
 
IndexOf won't work here as the available overloads for IndexOf on a span require the T to be IEquatable<T>, and in here T isn't constrained.